### PR TITLE
Add hook in select_common_type for CHAR/NCHAR case

### PR DESCRIPTION
**Pull request by wboettge**
### Description
Currently, babelfish gives an error when UNION is used with fixed-length types of different lengths or null. This change fixes how types and typmods are handled within UNION and VALUES via two new hooks. It also prevents the early-coercion of UNKNOWNs to handle NULL and string literals.
 
### Issues Resolved

BABEL-3348, BABEL-3392, BABEL-4157
 
### Check List

- [x] Commits are signed per the DCO using --signoff 

By submitting this pull request, I confirm that my contribution is under the terms of the PostgreSQL license, and grant any person obtaining a copy of the contribution permission to relicense all or a portion of my contribution to the PostgreSQL License solely to contribute all or a portion of my contribution to the PostgreSQL open source project.

For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/babelfish-for-postgresql/postgresql_modified_for_babelfish/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: src/backend/parser/analyze.c [hunk 1]
```diff
@@ -2278,8 +2279,8 @@ transformSetOperationTree(ParseState *pstate, SelectStmt *stmt,
 			if (lcoltype != UNKNOWNOID)
 				lcolnode = coerce_to_common_type(pstate, lcolnode,
 												 rescoltype, context);
-			else if (IsA(lcolnode, Const) ||
-					 IsA(lcolnode, Param))
```
Comment (Deepesh125) [inline src/backend/parser/analyze.c:2282]: why is this required?

File: src/backend/parser/parse_clause.c [hunk 2]
```diff
@@ -3294,11 +3294,12 @@ addTargetToSortList(ParseState *pstate, TargetEntry *tle,
 	/* if tlist item is an UNKNOWN literal, change it to TEXT */
 	if (restype == UNKNOWNOID)
 	{
-		tle->expr = (Expr *) coerce_type(pstate, (Node *) tle->expr,
```
Comment (Deepesh125) [inline src/backend/parser/parse_clause.c:3297]: what's meaning of this check? why is this required?

**Review by forestkeeper — APPROVED** [review]
LGTM
